### PR TITLE
PAPER - 11.1 CNN REPORT

### PR DESCRIPTION
**Pull request by JoseBrox**
Additions to paper:
1) "Convolutional neural network model for the implication graph" added as subsection 11.1. Also 1 small typo corrected in the intro to section 11
2) ref.bib updated with two new references
3) Fernando Vaquerizo-Villar added as coauthor

### CODE REVIEW

File: paper/ml.tex [hunk 1]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+
+\subsection{Convolutional neural network model for the implication graph}
+
+To model the implication graph, we used a convolutional neural network (CNN). For each pair of equations $(p,q)$, the input of the CNN consisted on a character-level tokenization (no vectorization) of the two equations, and the output of the CNN was a yes/no label depending on whether $p$ implies $q$. The CNN processed the input data using a 5-layer architecture, each layer composed of a 1-D convolution, followed by batch normalization and rectified linear unit activation functions (\cite{Goodfellow-et-al-2016}). After the last convolutional layer, a flattening layer and a softmax activation function were used to obtain the output of the network, i.e., the prediction of the implication for the input pair of equations. Note that we trained our models with the 16/10/2024 version of the (infinite) implication graph, for which 362 of the hardest implications (less than 0,002\% of the total) were still unknown.
```
Comment (dsm054) [inline paper/ml.tex:9]: consisted on -> of

Comment (dsm054) [inline paper/ml.tex:9]: I think we're using periods not commas for decimals.

File: paper/ml.tex [hunk 2]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+
+\smallskip
+
+Prior to training the CNN, we divided the data into training (60\%),validation (20\%), and test (20\%) subsets. The CNN was implemented in TensorFlow 2.9.0 (\cite{tensorflow2015-whitepaper}) and trained on an NVIDIA 3080 Ti GPU with the following configuration: the binary cross-entropy as the loss function to minimize, the Adam method with an initial learning rate of $10^{-3}$ for the adjustment of network weights, a batch size of 1024 with random shuffling, a learning rate reduction by a factor of 2 after 15 epochs without improvement in the validation loss, and early stopping if no improvement occurred for 40 epochs, being the model with the lowest validation loss retained as the final CNN.
```
Comment (dsm054) [inline paper/ml.tex:13]: This is a list of configs, but the last clause ("being the model.. retained") doesn't quite fit that pattern.

File: paper/ml.tex [hunk 3]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+The final CNN was evaluated on the test set, reaching prediction accuracy of 99.7\%, which means that the model mis-classified around 66k of the 22 million implications. Since this accuracy was somewhat surprising for such a small and “simple” model, as a control we generated a random label (yes/no) for each pair of equations, then trained the CNN on this data with 60\%/20\%/20\% training/validation/test percentages, resulting in a 49.99\% accuracy, as expected from a nonbiased model.
```
Comment (dsm054) [inline paper/ml.tex:14]: "reaching a" is slightly more fluid here.

Comment (dsm054) [inline paper/ml.tex:14]: misclassified, no need for hyphen

Comment (dsm054) [inline paper/ml.tex:14]: unbiased?  nonbiased sounds rare to my ear

File: paper/ml.tex [hunk 4]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+
+\smallskip
+
+It could be the case that the high accuracy of our CNN model was mostly due to it learning the transitivity of the implication relation, as opposed to it discovering patterns in the identities. To clarify this point, it was proposed to train our CNN model either on a random poset or on the equational poset with vertices and labels permuted, and check whether a similar accuracy was achieved. This experiment has not been performed yet.
+
+\smallskip
+
+In any case, since there are around 600k explicit implications from which the rest can be derived by transitivity (2.7\% of the total), if the CNN was learning transitivity it should perform well with a really small training dataset. Accordingly, we trained and assessed a CNN model with a 5\%/5\%/90\% training/validation/test proportion, with a resulting 99.6\% accuracy on test (the process finishing in under 20 minutes). But the high accuracy is maintained with even smaller training datasets, as evidenced in the following table:
```
Comment (dsm054) [inline paper/ml.tex:22]: "really"->"very"?  Also, maybe "on the test set"?

File: paper/ml.tex [hunk 5]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+  5/5/90 & 99.6 \\\hline
+  1/1/98 & 99.3 \\\hline
+  0.5/0.5/99 & 98.9 \\\hline
+  0.1/0.1/99.8 & 92.2 \\
+  \hline
+\end{tabular}
+\caption{Prediction accuracy as function of size of training set}
+\end{table}
+
+\noindent Since these training datasets were sensibly smaller than the subset of explicit implications, and were not carefully chosen from the poset extremes but taken randomly, we can conclude that even if the CNN was learning transitivity, probably this alone is not enough to explain the high accuracy achieved by the CNN model.
```
Comment (dsm054) [inline paper/ml.tex:39]: maybe reword "sensibly"?  If you're emphasizing the size, just use "much", or if you want to emphasize that the training datasets by nature had to be smaller than the implication dataset, maybe "naturally".  

Also, maybe "were learning", and "that alone is unlikely to" or "that by itself is probably insufficient" or something, adverbs of probability sometimes sound marked at the start of a clause.

File: paper/ml.tex [hunk 6]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+\smallskip
+
+Since sometimes machine learning is announced as a form of data compression, let us now comment on the level of data compression achieved by our CNN model. In the table below we compare the sizes of three different encodings of the implication graph: a) As the simplest approach, we can encode the full implication graph in one file as labelled pairs of equations of the form ($p$, $q$, yes/no). b) On the other extreme, we can encode it as a bit table not containing neither the explicit equation expressions nor their numbers, but just a 1/0 label for each point with coordinates ($p$,$q$), together with a table mapping each number to its corresponding equation expression and a small script to recover the file in a). c) Finally, we can encode it in the complete files of the CNN model produced by TensorFlow 2.9.0. In addition, we can either consider these files in their raw form, or we can highly (but losslessly) compress them to achieve a rough comparison of their actual information content; accordingly, in the table below we also include the sizes of the encoding models when compressed with 7-zip LZMA2 ultra compression with a 1536MB dictionary size and a 273 word size.
```
Comment (dsm054) [inline paper/ml.tex:43]: not containing neither -> containing neither

File: paper/ml.tex [hunk 7]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+\caption{Size of different encodings for the implication graph}
+\end{table}
+
+\smallskip
+
+\noindent As we see, the information in the CNN model is more than 13 times less than in the labelled pairs model, while it is more than 18 times that of the bit table model. We also note that the CNN model in its raw form is already quite incompressible, and more than 30 times smaller than the raw bit table.
+
+\smallskip
+
+Lastly, take also into account that our CNN model does not only encode the implication graph up to order 4 (with a 0.03\% of noise) but a priori may also be able to predict it for higher orders with a significant accuracy, and could thus be used to guide and speed up the determination of the implication graph up to order 5, by letting ATPs focus first on the CNN’s predicted status of the studied implication.
```
Comment (dsm054) [inline paper/ml.tex:65]: "also take into account" is better, although for some reason "take into account" is an unusual imperative in a paper.

Could also break the sentence at "and could thus", has lots of subparts now.

Comment (dsm054) [inline paper/ml.tex:65]: with a 0.03% -> can drop the "a"

File: paper/ml.tex [hunk 8]
```diff
@@ -2,7 +2,68 @@ \section{AI and Machine Learning Contributions}\label{ml-sec}
...
+\begin{tabular}{|c|c|c|}
+  \hline
+  % after \\: \hline or \cline{col1-col2} \cline{col3-col4} ...
+  \textbf{Encoding model} & \textbf{Uncompressed size} & \textbf{Compressed size} \\\hline\hline
+  Labelled pairs of equations & 1.5GB & 9MB \\\hline
+  Bit table & 42MB & 40KB \\\hline
+  CNN (99.7\% accuracy) & 1.34MB & 700KB \\
+  \hline
+\end{tabular}
+\caption{Size of different encodings for the implication graph}
```
Comment (dsm054) [inline paper/ml.tex:56]: sizes?

File: paper/references.bib [hunk 9]
```diff
@@ -914,3 +914,60 @@ @inproceedings{nieuwenhuis2005proof
   year={2005},
   organization={Springer}
 }
+
+@book{Goodfellow-et-al-2016,
+    title={Deep Learning},
+    author={Ian Goodfellow and Yoshua Bengio and Aaron Courville},
+    publisher={MIT Press},
+    doi= {10.1017/CBO9781139172752},
```
Comment (dsm054) [inline paper/references.bib:922]: Is this right?  Isn't this Term Rewriting?

https://www.cambridge.org/core/books/term-rewriting-and-all-that/71768055278D0DEF4FFC74722DE0D707

File: paper/references.bib [hunk 10]
```diff
@@ -914,3 +914,60 @@ @inproceedings{nieuwenhuis2005proof
...
+    url={\url{http://www.deeplearningbook.org}},
```
Comment (dsm054) [inline paper/references.bib:923]: do we need the \url?  If so, do we need it for tensorflow as well?

Comment (JoseBrox) [reply paper/references.bib]: We need the url because this format is what the authors want as citation (I'm not really sure if there is a DOI, the included one was just a mistake). The url for tensorflow is also included, aren't you able to see it?

**Comment by JoseBrox** [comment]
> Looks good to me; if we can resolve the comments above then I'm happy to merge.

Comments resolved! You can merge at any time.

File: paper/references.bib (hunk 10 above)
Comment (dsm054) [reply paper/references.bib]: I meant literally,`\url`, not url: we have `url={\url{http` here.  The Q is whether we need to put the `\url` marker around the url in tensorflow.
